### PR TITLE
[ISSUE-805] Upgrade from 1.0.x to 1.1.x

### PR DESCRIPTION
**Pull request by rkiselenko**
### Issue dell/csi-baremetal#805

Current upgrade flow broken due to OpenAPI validations in the deployments CRD.
As we know:

 _There is no support at this time for upgrading or deleting CRDs using[ Helm](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/)._

~~Perform CRD upgrade manually with `kubectl` to deliver newly version of CRDs to k8s.~~

Introduce a [`helm upgrade`](https://helm.sh/docs/helm/helm_upgrade/) hook with special job inside to perform CRDs upgrading.

Denote a rule - **all new created fields should have default values.**

## PR checklist
- [x] Add link to the issue
- [x] Choose Project
- [x] Choose PR label
- [ ] New unit tests added
- [ ] Modified code has meaningful comments
- [ ] All TODOs are linked with the issues
- [ ] All comments are resolved

## Testing

Requirements:

```bash
helm version
version.BuildInfo{Version:"v3.7.1", GitCommit:"1d11fcb5d3f3bf00dbe6fe31b8412839a96b3dc4", GitTreeState:"clean", GoVersion:"go1.16.9"}
kubectl version
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.3", GitCommit:"c92036820499fedefec0f847e2054d824aea6cd1", GitTreeState:"clean", BuildDate:"2021-10-27T18:41:28Z", GoVersion:"go1.16.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.11", GitCommit:"c6a2f08fc4378c5381dd948d9ad9d1080e3e6b33", GitTreeState:"clean", BuildDate:"2021-05-27T23:47:11Z", GoVersion:"go1.15.12", Compiler:"gc", Platform:"linux/amd64"}
WARNING: version difference between client (1.22) and server (1.19) exceeds the supported minor version skew of +/-1
```

### Goal 
All containers perform upgrade from 1.0.x to 1.1.x without or with minimal problems. 

### Prepare
In order to test our upgrade we use simple kind cluster:

```
make kind-create-cluster
kind create cluster --config test/kind/kind.yaml --image kindest/node:v1.19.11 --wait 30s
``` 

Install 1.0.x version on the test cluster with helm:

```bash
#!/bin/bash
set -e
export CSI_VERSION=1.0.0-482.4dbd5f3
export CSI_OPERATOR_VERSION=1.0.0-79.1e3fd76
export REGISTRY=registry
export DEPLOYMENT=artifactory
export OPERATOR=artifactory

helm install csi-baremetal-operator $OPERATOR \
        --kubeconfig /root/.kube/config \
        --set image.tag=$CSI_OPERATOR_VERSION \
        --set image.pullPolicy=IfNotPresent \
        --set global.registry=$REGISTRY
helm install csi-baremetal $DEPLOYMENT \
        --kubeconfig /root/.kube/config \
        --set image.tag=$CSI_VERSION \
        --set image.pullPolicy=IfNotPresent \
        --set global.registry=$REGISTRY \
        --set scheduler.patcher.enable=true \
        --set driver.drivemgr.type=loopbackmgr \
        --set scheduler.log.level=debug \
        --set nodeController.log.level=debug \
        --set driver.log.level=debug \
        --set scheduler.patcher.readinessTimeout=3 \
        --set driver.drivemgr.deployConfig=true
```

Now try to upgrade to 1.1.x with helm without path from this pull request:

```bash
#!/bin/bash
set -e
export CSI_VERSION=1.1.0-568.f9a3d59
export CSI_OPERATOR_VERSION=1.1.0-98.37e310c
export REGISTRY=registry
export DEPLOYMENT=artifactory
export OPERATOR=artifactory
helm upgrade csi-baremetal-operator $OPERATOR \
        --kubeconfig /root/.kube/config \
        --set image.tag=$CSI_OPERATOR_VERSION \
        --set image.pullPolicy=IfNotPresent \
        --set global.registry=$REGISTRY
helm upgrade csi-baremetal $DEPLOYMENT \
        --kubeconfig /root/.kube/config \
        --set image.tag=$CSI_VERSION \
        --set image.pullPolicy=IfNotPresent \
        --set global.registry=$REGISTRY \
        --set scheduler.patcher.enable=true \
        --set driver.drivemgr.type=loopbackmgr \
        --set scheduler.log.level=debug \
        --set nodeController.log.level=debug \
        --set driver.log.level=debug \
        --set scheduler.patcher.readinessTimeout=3 \
        --set driver.drivemgr.deployConfig=true
```
OpenAPI validation prevent us from upgrade:

```bash
Release "csi-baremetal-operator" has been upgraded. Happy Helming!
NAME: csi-baremetal-operator
LAST DEPLOYED: Mon Apr 11 13:02:10 2022
NAMESPACE: default
STATUS: deployed
REVISION: 2
TEST SUITE: None
Error: UPGRADE FAILED: error validating "": error validating data: [ValidationError(Deployment.spec.driver.controller): unknown field "fastDelay" in com.dell.csi-baremetal.v1.Deployment.spec.driver.controller, ValidationError(Deployment.spec.driver.controller): unknown field "maxFastAttempts" in com.dell.csi-baremetal.v1.Deployment.spec.driver.controller, ValidationError(Deployment.spec.driver.controller.sidecars.csi-provisioner): unknown field "args" in com.dell.csi-baremetal.v1.Deployment.spec.driver.controller.sidecars, ValidationError(Deployment.spec.driver.controller): unknown field "slowDelay" in com.dell.csi-baremetal.v1.Deployment.spec.driver.controller, ValidationError(Deployment.spec.driver.node): unknown field "serviceAccount" in com.dell.csi-baremetal.v1.Deployment.spec.driver.node, ValidationError(Deployment.spec.scheduler): unknown field "serviceAccount" in com.dell.csi-baremetal.v1.Deployment.spec.scheduler]
```
Validation fails due to missing fields in the CRDs. There is no `fastDelay` field in some object in our 1.0.x `CRD` and helm prevent us from success installation. Hence at this point we should have new CRD updated somehow before continue with `helm upgrade ...`. 

```bash
kubectl describe crd deployments.csi-baremetal.dell.com | grep Fast
```
No new fields (`Fast Delay`) 

> As workaround we can skip OpenAPI validation (add option `--disable-openapi-validation` to `helm upgrade ...`.

Lets apply new CRD and perform `heml upgrade ...` with new versions.

```bash
export CSI_OPERATOR_VERSION=VERSION WITH PATH
wget http://artifactory/csi-operator/$CSI_OPERATOR_VERSION/csi-baremetal-operator-$CSI_OPERATOR_VERSION.tgz
tar -xzvf csi-baremetal-operator-$CSI_OPERATOR_VERSION.tgz
kubectl apply -f csi-baremetal-operator/crds/
```
Now try to upgrade to 1.1.x with helm with path from this pull request:

```bash
#!/bin/bash
set -e
export CSI_VERSION=1.1.0-567.245c767
export CSI_OPERATOR_VERSION=1.1.0-99.a0d9afb
export REGISTRY=registry
export DEPLOYMENT=artifactory
export OPERATOR=artifactory
helm upgrade csi-baremetal-operator $OPERATOR \
        --kubeconfig /root/.kube/config \
        --set image.tag=$CSI_OPERATOR_VERSION \
        --set image.pullPolicy=IfNotPresent \
        --set global.registry=$REGISTRY
helm upgrade csi-baremetal $DEPLOYMENT \
        --kubeconfig /root/.kube/config \
        --set image.tag=$CSI_VERSION \
        --set image.pullPolicy=IfNotPresent \
        --set global.registry=$REGISTRY \
        --set scheduler.patcher.enable=true \
        --set driver.drivemgr.type=loopbackmgr \
        --set scheduler.log.level=debug \
        --set nodeController.log.level=debug \
        --set driver.log.level=debug \
        --set scheduler.patcher.readinessTimeout=3 \
        --set driver.drivemgr.deployConfig=true
```

All upgrade done without any problem with validations:

```bash
Release "csi-baremetal-operator" has been upgraded. Happy Helming!
NAME: csi-baremetal-operator
LAST DEPLOYED: Mon Apr 11 13:28:11 2022
NAMESPACE: default
STATUS: deployed
REVISION: 2
TEST SUITE: None
Release "csi-baremetal" has been upgraded. Happy Helming!
NAME: csi-baremetal
LAST DEPLOYED: Mon Apr 11 13:28:12 2022
NAMESPACE: default
STATUS: deployed
REVISION: 2
TEST SUITE: None
NOTES:
This CSI plugin is a part of platform Atlantic.
It is used to provide dynamic volume provisioning on bare-metal k8s cluster.
To create Persistent Volume for a pod it's needed to set Storage Class csi-baremetal-sc
in corresponding Persistent Volume Claim.

Your release is named csi-baremetal.
```

All new fields already updated and our upgrade done without any problem.

```bash
kubectl describe crd deployments.csi-baremetal.dell.com | grep Fast
                      Fast Delay:
                        Description:  FastDelay is the parameter for NewItemFastSlowRateLimiter in Reservation Controller
                      Max Fast Attempts:
                        Description:  MaxFastAttempts is the parameter for NewItemFastSlowRateLimiter in Reservation Controller
                        Description:  SlowDelay is the parameter for NewItemFastSlowRateLimiter in Reservation Controller
```

### CODE REVIEW

**Review by mitsaa — APPROVED** [review]

File: docs/README.md [hunk 1]
```diff
@@ -120,6 +120,42 @@ Installation process
...
+    export CSI_OPERATOR_VERSION=VERSION WITH PATH
+    wget http://artifactory/csi-operator/$CSI_OPERATOR_VERSION/csi-baremetal-operator-$CSI_OPERATOR_VERSION.tgz
+    tar -xzvf csi-baremetal-operator-$CSI_OPERATOR_VERSION.tgz
+    kubectl apply -f csi-baremetal-operator/crds/
+    ```
+  * Upgrade CSI Operator
+      ```shell script
+      helm upgrade csi-baremetal-operator csi/csi-baremetal-operator --set global.registry=$REGISTRY \
+    --set global.registrySecret=$DOCKER_REGISTRY_SECRET
+      ```
```
Comment (safronovD) [inline docs/README.md:158]: We already have **Upgrade process** section below in this doc

**Review by safronovD — APPROVED** [review]

File: docs/README.md [hunk 2]
```diff
@@ -120,6 +120,18 @@ Installation process
       helm install csi-baremetal csi/csi-baremetal-deployment --set driver.drivemgr.type=halmgr \
       --set global.registry=$REGISTRY --set global.registrySecret=$DOCKER_REGISTRY_SECRET
       ```
+* Note about Upgrade
```
Comment (mishoyama) [inline docs/README.md:123]: Please put in **Upgrade process** section. Should it be done before or after `helm upgrade`?

**Review by mishoyama — CHANGES_REQUESTED** [review]

File: docs/README.md (hunk 2 above)
Comment (rkiselenko) [reply docs/README.md]: > Should it be done before or after helm upgrade?

No matter before or after, it must be done if CRDs update expected

File: docs/README.md [hunk 3]
```diff
@@ -236,6 +236,17 @@ Usage
...
 ---------------------
 To upgrade please reference _Installation process_ section but replace `helm install` by `helm upgrade` command
+
+* Note about Upgrade
+  > There is no support at this time for upgrading or deleting CRDs using [Helm](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/).
+
+  In order to upgrade CRD use `kubectl apply -f ...` command.
+  ```bash
+  export CSI_OPERATOR_VERSION=VERSION WITH PATH
+  wget http://artifactory/csi-operator/$CSI_OPERATOR_VERSION/csi-baremetal-operator-$CSI_OPERATOR_VERSION.tgz
```
Comment (mishoyama) [inline docs/README.md:246]: BTW what do you mean by `artifactory`?

Comment (mishoyama) [reply docs/README.md:246]: we need to have instructions which can be simply copied and executed

Comment (rkiselenko) [reply docs/README.md:246]: It's depends on requirements, there is charts placed. The goal is download charts and apply.

**Review by eddiepavkovic — COMMENTED** [review]
Can we make so that it uses a pre-upgrade hook like we do for ObjectScale.

create a pre-upgrade helm hook then use a docker image
https://eos2git.cec.lab.emc.com/ECS/objectscale-charts/pull/23/files

File: docs/README.md (hunk 3 above)
Comment (eddiepavkovic) [reply docs/README.md]: Or do what we did for objectscale and make it automatic: 
https://eos2git.cec.lab.emc.com/ECS/objectscale-charts/pull/23/files

Comment (mishoyama) [reply docs/README.md]: @eddiepavkovic please don't add links to internal dell resources in public repo. Thanks!

File: build/CRD.Dockerfile [hunk 4]
```diff
@@ -0,0 +1,14 @@
+#
```
Comment (mishoyama) [inline build/CRD.Dockerfile:1]: build folder should not contain any code. Please create separate folder similar to other components

File: charts/csi-baremetal-operator/templates/pre-upgrade-crds.yaml [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    "helm.sh/hook-delete-policy": before-hook-creation,hook-succeeded,hook-failed
+rules:
+  - apiGroups:
+      - apiextensions.k8s.io
+    resources:
+      - customresourcedefinitions
+    verbs:
+      - '*'
+  - apiGroups:
+      - ""
```
Comment (mishoyama) [inline charts/csi-baremetal-operator/templates/pre-upgrade-crds.yaml:60]: No need to have all

File: charts/csi-baremetal-operator/templates/pre-upgrade-crds.yaml [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    resources:
+      - serviceaccounts
+    verbs:
+      - get
```
Comment (mishoyama) [inline charts/csi-baremetal-operator/templates/pre-upgrade-crds.yaml:64]: May be only replace?

File: charts/csi-baremetal-operator/templates/pre-upgrade-crds.yaml [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+{{- if .Values.global.registrySecret }}
+imagePullSecrets:
+  - name: {{ .Values.global.registrySecret }}
+{{- end }}
+
+#
+# Create the cluster role for replacing CRDs during upgrade.
+#
+---
+kind: ClusterRole
```
Comment (mishoyama) [inline charts/csi-baremetal-operator/templates/pre-upgrade-crds.yaml:39]: We need to delete all roles on post upgrade hook

File: charts/csi-baremetal-operator/templates/pre-upgrade-crds.yaml [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+      imagePullSecrets:
+        - name: {{ .Values.global.registrySecret }}
+      {{- end }}
+      containers:
+        - name: "{{.Release.Namespace}}-{{.Release.Name}}-csi-baremetal-pre-crds-job"
+          image: {{ default .Values.global.registry .Values.preUpgradeCRDsHooks.registry}}/{{ .Values.preUpgradeCRDsHooks.repository }}:{{ .Values.preUpgradeCRDsHooks.version }}
+          imagePullPolicy: Always
+          command: ["/bin/sh", "-c"]
+          args:
+          - kubectl -n {{ .Release.Namespace }} replace crd -f ./crds/{{ .Chart.Name }}/*;
```
Comment (safronovD) [inline charts/csi-baremetal-operator/templates/pre-upgrade-crds.yaml:131]: CRD is cluster scoped. Do we need to set namespace here?

File: build/CRD.Dockerfile (hunk 4 above)
Comment (safronovD) [reply build/CRD.Dockerfile]: done

File: charts/csi-baremetal-operator/templates/pre-upgrade-crds.yaml (hunk 5 above)
Comment (safronovD) [reply charts/csi-baremetal-operator/templates/pre-upgrade-crds.yaml]: It's not working. `kubectl replace` call contains update, delete, get and list requests

File: charts/csi-baremetal-operator/templates/pre-upgrade-crds.yaml (hunk 6 above)
Comment (safronovD) [reply charts/csi-baremetal-operator/templates/pre-upgrade-crds.yaml]: SA removed

File: charts/csi-baremetal-operator/templates/pre-upgrade-crds.yaml (hunk 7 above)
Comment (safronovD) [reply charts/csi-baremetal-operator/templates/pre-upgrade-crds.yaml:35]: Resource deletion strategy contains in the helm-based annotation
```
"helm.sh/hook-delete-policy": before-hook-creation,hook-succeeded,hook-failed
```

It's not need to have separate functional

File: charts/csi-baremetal-operator/templates/pre-upgrade-crds.yaml (hunk 8 above)
Comment (safronovD) [reply charts/csi-baremetal-operator/templates/pre-upgrade-crds.yaml]: done

File: docs/README.md [hunk 9]
```diff
@@ -242,7 +242,31 @@ Usage
...
 ---------------------
-To upgrade please reference _Installation process_ section but replace `helm install` by `helm upgrade` command
+
+###  Automated
+
+To upgrade please reference Installation process section but replace `helm install` by `helm upgrade` command
+
+See [helm upgrade](https://helm.sh/docs/helm/helm_upgrade/) for command documentation.
+
+### Manual
```
Comment (mishoyama) [inline docs/README.md:252]: Do we really need this part? If it failed we need to fix problem

File: hook/CRD.Dockerfile [hunk 10]
```diff
@@ -0,0 +1,18 @@
+#
```
Comment (mishoyama) [inline hook/CRD.Dockerfile:1]: Can it be just Dockerfile without CRD?

**Review by misanthropicat — APPROVED** [review]

File: docs/README.md (hunk 9 above)
Comment (safronovD) [reply docs/README.md]: Ok

**Review by mishoyama — APPROVED** [review]

**Pull request merged**